packages/ai-jsx/src/batteries/use-tools.tsx

```diff
@@ -246,25 +256,40 @@ export async function* UseToolsFunctionCall(props: UseToolsProps, { render }: Re
     }
 
     if (functionCallElement) {
-      messages.push(functionCallElement);
-      yield functionCallElement;
       // Call the selected function and append the result to the messages.
       let response;
       try {
         const callable = props.tools[functionCallElement.props.name].func;
         response = await callable(functionCallElement.props.args);
       } catch (e: any) {
-        response = `Function called failed with error: ${e.message}.`;
+        response = `Function call to ${functionCallElement.props.name} failed with error: ${e.message}.`;
       } finally {
-        const functionResponse = <FunctionResponse name={functionCallElement.props.name}>{response}</FunctionResponse>;
-        yield functionResponse;
-        messages.push(functionResponse);
+        const functionResponse = memo(
+          <FunctionResponse name={functionCallElement.props.name}>{response}</FunctionResponse>
+        );
+        if (props.showSteps) {
+          yield (
+            <>
+              {'\n'}
+              {functionResponse}
+              {'\n'}
+            </>
```

Comment:
Hmm, seems like a "presentation" issue. Maybe we should have a better way of separating the "presentation" from content (I have the same issue in multiple places), but I understand why you put this here and it's probably better to have it. Still, you can't remove the `\n`'s if you don't want them which is not good.

Reply:
Generally I have this "presentation" issue in multiple places (e.g. Image and Audio tags). `map` is nice for modifying results, but I can't recall why I don't like it as much. Does `map` only work on the intermediate results?